rust/plugin/src/godot_project.rs

```diff
@@ -542,56 +536,60 @@ impl GodotProject {
 
     fn spawn_sync_task(
         runtime: &Runtime,
-        docs_state: Arc<Mutex<HashMap<DocumentId, Automerge>>>,
-        doc_handles_state: Arc<Mutex<HashMap<DocumentId, DocHandle>>>,
+        mut new_docs: futures::channel::mpsc::UnboundedReceiver<DocHandle>,
+        doc_handles_state: DocHandleMap,
         sync_event_sender: Sender<SyncEvent>,
     ) {
-        let docs_state_clone = docs_state.clone();
-        let doc_handles_state_clone = doc_handles_state.clone();
+        let initial_handles = doc_handles_state
+            .current_handles();
 
         runtime.spawn(async move {
-            loop {
-                let mut futures = FuturesUnordered::new();
+            // This is a stream of (DocumentId, SyncEvent)
+            let mut all_doc_changes = futures::stream::SelectAll::new();
+
+            // First add a stream for all the initial documents to the SelectAll
+            for doc_handle in initial_handles {
+                let doc_id = doc_handle.document_id();
+                let doc_handle = doc_handle.clone();
+
+                let change_stream = handle_changes(doc_handle.clone()).filter_map(move |diff| {
+                    let doc_id = doc_id.clone();
+                    async move {
+                        if diff.is_empty() {
+                            None
+                        } else {
+                            Some(SyncEvent::DocChanged { doc_id: doc_id.clone() } )
+                        }
+                    }
+                });
 
-                // Create initial futures for all doc handles
-                let doc_handles: Vec<DocHandle> = {
-                    let handles = doc_handles_state_clone.lock().unwrap();
-                    handles.values().cloned().collect()
-                };
+                all_doc_changes.push(change_stream.boxed());
+            }
 
-                for doc_handle in doc_handles.iter() {
-                    let doc_handle = doc_handle.clone();
-                    let future: Pin<Box<dyn Future<Output = DocHandle> + Send>> =
-                        Box::pin(async move {
-                            doc_handle.changed().await.unwrap();
-                            doc_handle
+            // Now, drive the SelectAll and also wait for any new documents to  arrive and add
+            // them to the selectall
+            loop {
+                futures::select! {
+                    sync_event = all_doc_changes.select_next_some() => {
+                        sync_event_sender.send(sync_event).unwrap();
```

Comment:
Probably we should change `sync_event_sender` to a `futures::channel::mpsc::UnboundedSender` but I wanted to make this a minimal PR.